### PR TITLE
[serve] Add initial health check before marking a replica as RUNNING

### LINKED ISSUE #26204: [serve] Replica failing health check doesn't cause REST API to return unhealthy
`tmp.py`:
```python
from ray import serve

from fastapi import FastAPI


@serve.deployment
class Model:
    def __init__(self, arg):
        self._arg = arg

    def check_health(self):
        print(self._arg)
        if self._arg == "unhealthy":
            raise Exception("hihihi")

    def reconfigure(self, c):
        self._arg = c["arg"]

    def __call__(self, *args):
        return self._arg


app = FastAPI()


@serve.deployment
@serve.ingress(app)
class Driver:
    def __init__(self, models):
        self._models = models

    @app.get("/{model}")
    async def hi(self, model: str):
        return await self._models[model].remote()


app = Driver.bind({"a": Model.bind("a"), "b": Model.bind("b")})
```

`app.yaml`:
```
# This file was generated using the `serve build` command on Ray v3.0.0.dev0.

import_path: tmp:app

runtime_env: {}

deployments:

- name: Model
  num_replicas: 1
  route_prefix: null
  max_concurrent_queries: 100
  user_config: {"arg": "unhealthy"}
  autoscaling_config: null
  graceful_shutdown_wait_loop_s: 2.0
  graceful_shutdown_timeout_s: 20.0
  health_check_period_s: 10.0
  health_check_timeout_s: 30.0
  ray_actor_options: null

- name: Model_1
  num_replicas: 1
  route_prefix: null
  max_concurrent_queries: 100
  user_config: null
  autoscaling_config: null
  graceful_shutdown_wait_loop_s: 2.0
  graceful_shutdown_timeout_s: 20.0
  health_check_period_s: 10.0
  health_check_timeout_s: 30.0
  ray_actor_options: null

- name: Driver
  num_replicas: 1
  route_prefix: /
  max_concurrent_queries: 100
  user_config: null
  autoscaling_config: null
  graceful_shutdown_wait_loop_s: 2.0
  graceful_shutdown_timeout_s: 20.0
  health_check_period_s: 10.0
  health_check_timeout_s: 30.0
  ray_actor_options: null
```

Status (despite controller logs saying it's being marked unhealthy):
```
app_status:
  status: RUNNING
  message: ''
  deployment_timestamp: 1656537678.8220658
deployment_statuses:
- name: Model
  status: HEALTHY
  message: ''
- name: Model_1
  status: HEALTHY
  message: ''
- name: Driver
  status: HEALTHY
  message: ''
```

### PR DESCRIPTION
**Pull request by zcin**
Signed-off-by: Cindy Zhang <cindyzyx9@gmail.com>

## Why are these changes needed?

A deployment is marked as HEALTHY the moment we have reached the target number of RUNNING replicas. However, this is deceiving when replicas are repeatedly failing health checks, because the replicas will be marked RUNNING after successful initialization, and the deployment status will not change until 30 seconds later when the replica fails 3 consecutive health checks and is restarted (then the loop starts again).

This change will ensure that a replica is not "healthy by default"; instead, the replica must pass one health check immediately after startup. If it does, the replica is marked as RUNNING; if not, it is treated like a startup failure and stopped and restarted on the spot.

## Related issue number

"Closes #26204"

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by zcin** [comment]
Windows tests unrelated.

File: python/ray/serve/_private/deployment_state.py [hunk 1]
```diff
@@ -69,8 +69,9 @@ class ReplicaState(Enum):
 class ReplicaStartupStatus(Enum):
     PENDING_ALLOCATION = 1
     PENDING_INITIALIZATION = 2
-    SUCCEEDED = 3
-    FAILED = 4
+    PENDING_INITIAL_HEALTH_CHECK = 3
```
Comment (edoakes) [inline python/ray/serve/_private/deployment_state.py:72]: what's the reason to separate this from `PENDING_INITIALIZATION`? if there is no need for the distinction let's err on the side of having fewer states (easier to reason about)

**Review by edoakes — COMMENTED** [review]
I wonder if it's possible to totally unify this with the initialization check -- can we just make the health check call first check that the user class is initialized and use the same method for both? Basically in `replica.py`:

```
async def check_health_wrapper(self):
    await self._wait_for_constructor()
    ... do health check as we do now ...

File: python/ray/serve/_private/deployment_state.py [hunk 2]
```diff
@@ -459,23 +460,43 @@ def check_ready(self) -> Tuple[ReplicaStartupStatus, Optional[DeploymentVersion]
...
-                self._graceful_shutdown_timeout_s = (
-                    deployment_config.graceful_shutdown_timeout_s
-                )
-                self._health_check_period_s = deployment_config.health_check_period_s
-                self._health_check_timeout_s = deployment_config.health_check_timeout_s
-                self._node_id = ray.get(self._allocated_obj_ref).hex()
-            except Exception:
-                logger.exception(f"Exception in deployment '{self._deployment_name}'")
-                return ReplicaStartupStatus.FAILED, None
+
```
Comment (sihanwang41) [inline python/ray/serve/_private/deployment_state.py:463]: please update the `check_ready` method description.

File: python/ray/serve/tests/test_healthcheck.py [hunk 3]
```diff
@@ -202,6 +203,27 @@ def check_fails_3_times():
     check_fails_3_times()
 
 
+def test_health_check_failure_makes_deployment_unhealthy(serve_instance):
```
Comment (sihanwang41) [inline python/ray/serve/tests/test_healthcheck.py:206]: add a test:
The deployment is `HEALTHY` -> `UNHEALTHY` when there is a replica failing the health check.

File: python/ray/serve/_private/replica.py [hunk 4]
```diff
@@ -214,13 +214,25 @@ async def is_allocated(self) -> str:
...
+
+        async def reconfigure(
+            self, user_config: Optional[Any] = None
         ) -> Tuple[DeploymentConfig, DeploymentVersion]:
             # Unused `_after` argument is for scheduling: passing an ObjectRef
             # allows delaying reconfiguration until after this call has returned.
-            if self.replica is None:
-                await self._initialize_replica()
             if user_config is not None:
                 await self.replica.reconfigure(user_config)
```
Comment (edoakes) [inline python/ray/serve/_private/replica.py:237]: this is a nice simplification :)

Comment (edoakes) [reply python/ray/serve/_private/replica.py:237]: nit: we seem to be mixing the terminology of "ready" and "initialized" -- let's pick one and standardize on it both here and in the controller?

File: python/ray/serve/tests/test_healthcheck.py [hunk 5]
```diff
@@ -202,6 +202,78 @@ def check_fails_3_times():
...
+        return (
+            app_status.deployment_statuses[0].name == "WillBeUnhealthy"
+            and app_status.deployment_statuses[0].status == expected_status
+        )
+
+    toggle = ray.remote(Toggle).remote()
+    serve.run(WillBeUnhealthy.bind(toggle))
+
+    # Check that deployment is healthy initially
+    assert check_status("HEALTHY")
```
Comment (edoakes) [inline python/ray/serve/tests/test_healthcheck.py:265]: nit: use enum instead of string directly

**Review by edoakes — APPROVED** [review]

File: python/ray/serve/tests/test_healthcheck.py (hunk 3 above)
Comment (zcin) [reply python/ray/serve/tests/test_healthcheck.py:206]: Done!

**Comment by zcin** [comment]
@edoakes @sihanwang41 Addressed comments, PTAL!

**Comment by scv119** [comment]
@zcin this breaks bk://:java and bk://:mac: :apple: Ray C++ and Java 
image

**Comment by edoakes** [comment]
That's on me, wasn't careful enough with merging. @zcin could you make a revert PR?

**Pull request merged**